### PR TITLE
Updated File Management for Experiments

### PR DESCRIPTION
**Pull request by mshalm**
As referenced in #6 and #7 , The previous file system management made storing files associated with individual experiments and learned systems difficult. Required capabilities include:

- A filename for a Drake `VideoWriter` to store trajectory visualizations at for `DrakeSystem`s.
- A folder to export learned URDFs to for `MultibodyLearnableSystem`s.
- Easy access to the above for `DrakeExperiment`s -- not just `DrakeMultibodyLearnableExperiment`s -- to do an overlayed visualization for ground-truthand learned geometries.
- Various use cases for a canonical experiment name.

We provide this functionality with the following changes:
- All `SupervisedLearningExperiment`s have a unique `name` parameter available in their `SupervisedLearningExperimentConfig`.
-  All `SupervisedLearningExperiment`s produce a "joint" `SystemSummary` during training, which produces scalars, videos, and/or meshes that capture information that compares the experiment's base system to it's learned system. By default, this summary is empty.
- the joint `SystemSummary` for `DrakeExperiment`s produces the overlayed trajectory. As part of this, the doubled-URDF `DrakeSystem` that used to be generated as part of `MultibodyLearnableSystem.get_visualization_system` has been moved to the `DrakeExperiment` class, as really this system jointly depends on both the base `DrakeSystem` and learned system. This generation by default copies the base system's geometries for visualization, and will run seamlessly with an end-to-end `DeepLearnableSystem`s as well as `MultibodyLearnableSystem`s.
- The output URDF directory of a `MultibodyLearnableSystem` is now an optional parameter for it's constructor.
- `MultibodyPlantDiagram` and `DrakeSystem` now also take an optional `visualization_file` parameter to store trajectory visualizations at, and add a `VideoWriter` to the system if and only if this parameter is set.
- `file_utils.py` has been changed to reflect the above.

Various tweaks to static typing and code style were implemented to pass `mypy` and `pylint` tests.

### CODE REVIEW

File: dair_pll/drake_experiment.py [hunk 1]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
 class MultibodyLearnableSystemConfig(DrakeSystemConfig):
     loss: MultibodyLosses = MultibodyLosses.PREDICTION_LOSS
     """Whether to use ContactNets or prediction loss."""
+    urdfs_output_directory: str = "./"
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: Could be useful to write what this path is relative to.

File: dair_pll/drake_experiment.py [hunk 2]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+    """Where to store learned URDFs"""
 
 
-class DrakeExperiment(SupervisedLearningExperiment):
-    drake_system: Optional[DrakeSystem]
+@dataclass
+class DrakeExperimentConfig(SupervisedLearningExperimentConfig):
+    joint_visualization_file: str = "./{TRAJECTORY_GIF_DEFAULT_NAME}"
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: Does this need to be an f-string for the result to be what you're expecting?

File: dair_pll/drake_experiment.py [hunk 3]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+
+        Implemented as a thin wrapper of
+        ``vis_utils.generate_visualization_system()``, which generates a
+        drake system where each model in the base
+        :py:class:`DrakeSystem` has a duplicate, and visualization
+        elements are repainted for visual distinction.
+
+        Args:
+            visualization_file: Output GIF filename for trajectory video.
+            new_geometry: Whether to use learned geometry.
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: Doesn't look like `new_geometry` is actually an argument for this method.

File: dair_pll/drake_experiment.py [hunk 4]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+
+        Args:
+            learned_system: System being learned in experiment.
+
+        Returns:
+            Drake version of learned system.
+        """
+        return None
+
+    def do_force_visualizer_regeneration(self) -> bool:
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: This could benefit from a typical method/function description.  Upon first reading of the function name, I assumed this was something to do with visualizing forces, which is not what this actually is used for.

File: dair_pll/drake_experiment.py [hunk 5]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+
+        Returns:
+            New :py:class:`DrakeSystem` with doubled state and repainted
+            elements.
+        """
+        # Generate a new visualization system if it needs to use the updated
+        # geometry, or if it hasn't been created yet.
+        force_regeneration = self.do_force_visualizer_regeneration()
+        if force_regeneration or self.visualization_system is None:
+            # learned_system = None
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: Any reason these commented out lines are still lingering?

File: dair_pll/drake_experiment.py [hunk 6]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+        e.g. data.
+
+        Implemented as a thin wrapper of
+        ``vis_utils.generate_visualization_system()``, which generates a
+        drake system where each model in the base
+        :py:class:`DrakeSystem` has a duplicate, and visualization
+        elements are repainted for visual distinction.
+
+        Args:
+            visualization_file: Output GIF filename for trajectory video.
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: Is there a particular reason for `visualization_file` being an explicit argument to this method?  Seems like the method could just define `visualization_file = cast(DrakeExperimentConfig, self.config).joint_visualization_file` for itself, requiring `learned_system` as the only argument.

File: dair_pll/drake_experiment.py [hunk 7]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+            self.visualization_system = \
+                vis_utils.generate_visualization_system(
+                    base_system,
+                    visualization_file,
+                    learned_system=self.get_learned_drake_system(learned_system)
+                )
+
+        return self.visualization_system
+
+    def joint_system_summary(self, statistics: Dict, learned_system: System) \
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py]: Class method comment description?

File: dair_pll/drake_experiment.py [hunk 8]
```diff
@@ -30,44 +32,166 @@ class MultibodyLosses(Enum):
...
+                    statistics[prediction_key][traj_num])
+                visualization_trajectory = torch.cat(
+                    (space.q(target_trajectory), space.q(prediction_trajectory),
+                     space.v(target_trajectory),
+                     space.v(prediction_trajectory)), -1)
+                video, framerate = vis_utils.visualize_trajectory(
+                    visualization_system, visualization_trajectory)
+                videos[f'{set_name}_trajectory_prediction_{traj_num}'] = \
+                    (video, framerate)
+        return SystemSummary(scalars={}, videos=videos, meshes={})
```
Comment (ebianchi) [inline dair_pll/drake_experiment.py:160]: Perhaps this could be clarified in a comment header for this class method, but it's not clear to me why this would return a system summary only filled with video reporting, with scalars and meshes empty.

File: dair_pll/drake_utils.py [hunk 9]
```diff
@@ -251,15 +268,18 @@ def __init__(self,
...
+                                                  fps=FPS,
+                                                  width=VIDEO_PIXELS[1],
+                                                  height=VIDEO_PIXELS[0],
+                                                  fov_y=CAM_FOV)
 
         # Adds ground plane at ``z = 0``
-        halfspace_transform = RigidTransform()
-        friction = CoulombFriction(1.0, 1.0)
+        halfspace_transform = RigidTransform_[float]()
+        friction = CoulombFriction_[float](1.0, 1.0)
```
Comment (ebianchi) [inline dair_pll/drake_utils.py:282]: Just out of curiosity, why these changes from `RigidTransform, CoulombFriction` to `RigidTransform_[float], CoulombFriction_[float]`?

File: dair_pll/multibody_learnable_system.py [hunk 10]
```diff
@@ -339,20 +275,19 @@ def forward_dynamics(self,
...
-        for i in range(n_contacts):
-            reorder_mat[i][3 * i + 2] = 1
-            reorder_mat[n_contacts + 2 * i][3 * i] = 1
-            reorder_mat[n_contacts + 2 * i + 1][3 * i + 1] = 1
-        reorder_mat = reorder_mat.reshape(
-            (1,) * (delassus.dim() - 2) + reorder_mat.shape).expand(
-            delassus.shape)
-
-
+        n_force = n_contacts * 3
```
Comment (ebianchi) [inline dair_pll/multibody_learnable_system.py]: This comment isn't due to new changes, but an artifact of the original code:  This `reorder_mat` construction is repeated in two places.  Might be useful to wrap it into one helper function and call just that in two places.

File: dair_pll/multibody_learnable_system.py [hunk 11]
```diff
@@ -388,41 +326,22 @@ def sim_step(self, x: Tensor, carry: Tensor) -> Tuple[Tensor, Tensor]:
...
+        :py:class:`~dair_pll.geometry.DeepSupportConvex` es.
 
         Args:
             statistics: Updated evaluation statistics for the model.
-            new_geometry: Whether to use new learned geometry or the original
-              URDF's geometry in the rollout videos.
 
         Returns:
-            Scalars and videos packaged into a ``SystemSummary``.
+            Scalars and meshes packaged into a ``SystemSummary``.
```
Comment (ebianchi) [inline dair_pll/multibody_learnable_system.py:329]: Before these changes, this `summary()` method actually did create a complete `SystemSummary`, full of meshes, scalars, and videos.  Now, this function does scalars and meshes, while `DrakeExperiment.joint_system_summary()` does videos, so my guess is that for a complete summary both of these methods need to be called and the results combined.  It could be helpful to add some comments that speak to how this incomplete `summary()` relates to other methods that also do partial summaries.  Maybe different naming choices could make this more straightforward.

File: dair_pll/vis_utils.py [hunk 12]
```diff
@@ -134,9 +136,16 @@ def visualize_trajectory(drake_system: DrakeSystem,
         (1, T, 3, H, W) ndarray video capture of trajectory with resolution
         H x W, which are set to 480x640 in :py:mod:`dair_pll.drake_utils`.
         The true framerate, rounded to an integer.
+
+    Todo:
```
Comment (ebianchi) [inline dair_pll/vis_utils.py:140]: Helpful note that should have been there from before, thanks!

File: examples/contactnets_simple.py
Comment (ebianchi) [inline examples/contactnets_simple.py:1]: Line 218 `learned_system.generate_updated_urdfs(storage_name)` needs to be updated to just `learned_system.generate_updated_urdfs()` with these changes.  Otherwise running this test script with `--regenerate` fails.

File: examples/contactnets_simple.py [hunk 13]
```diff
@@ -42,8 +42,11 @@
 ELBOW_URDFS = {MESH_TYPE: ELBOW_MESH_URDF_ASSET, BOX_TYPE: ELBOW_BOX_URDF_ASSET}
 URDFS = {CUBE_SYSTEM: CUBE_URDFS, ELBOW_SYSTEM: ELBOW_URDFS}
 
-STORAGE_NAME = os.path.join(os.path.dirname(__file__), 'storage',
-                            CUBE_DATA_ASSET)
+DATA_STORAGE_NAME = os.path.join(os.path.dirname(__file__), 'storage',
```
Comment (ebianchi) [inline examples/contactnets_simple.py]: `DATA_STORAGE_NAME` and `RESULTS_STORAGE_NAME` aren't being used in this test script.

File: examples/contactnets_simple.py [hunk 14]
```diff
@@ -110,7 +113,12 @@ def main(system: str = CUBE_SYSTEM,
     data_asset = DATA_ASSETS[system]
     storage_name = os.path.join(os.path.dirname(__file__), 'storage',
                                 data_asset)
+
```
Comment (ebianchi) [inline examples/contactnets_simple.py:114]: Re the above comment:  in favor of the definitions of `DATA_STORAGE_NAME` and `RESULTS_STORAGE_NAME`, looks like you can just remove lines 114 through 118, which re-implement those definitions, and use those constants instead.

File: dair_pll/file_utils.py
Comment (ebianchi) [inline dair_pll/file_utils.py:1]: The current file structure for an experiment, say named `exp`, is:
```
- results/
  |
  - exp/
    |
    - urdfs/
    - some files
- storage/
  |
  - exp/
    |
    - data/
    - temp/
    - tensorboard/
```

I find the following to be better:
```
- exp/
  |
  - storage/
    |
    - data/
    - temp/
    - tensorboard/
  - results/
    |
    - urdfs/
    - some files
```
...or:
```
- exp/
  |
  - data/
  - temp/
  - tensorboard/
  - urdfs/            # this could instead be inside exp/results/
  - results/
    |
    - some files
```
...which avoids creating two `exp` folders in different places.  Is there any reason that creating two folders instead of one is better?

**Review by ebianchi — CHANGES_REQUESTED** [review]
See in-line comments for more specific comments/questions/requests.  The larger-picture question/concern I had was about how the proposed implementation has experiment setup and results split into two separate locations.  It seems more convenient for there to be one folder for a given experiment name, and the folder contains both setup files (e.g. data, tensorboard events, etc.) and results (e.g. generated URDFs, temporary gifs, etc.).  Feel free to provide reasons for the proposed dual-location setup instead, as perhaps I've overlooked some benefit of that approach.

File: dair_pll/drake_experiment.py (hunk 1 above)
Comment (mshalm) [reply dair_pll/drake_experiment.py]: In the docstring you mean?

Comment (ebianchi) [reply dair_pll/drake_experiment.py]: Yes

File: dair_pll/drake_experiment.py (hunk 3 above)
Comment (mshalm) [reply dair_pll/drake_experiment.py]: Agreed, missed that this should have been eliminated from the comment.

File: dair_pll/drake_experiment.py (hunk 2 above)
Comment (mshalm) [reply dair_pll/drake_experiment.py]: agreed

File: dair_pll/drake_experiment.py (hunk 4 above)
Comment (mshalm) [reply dair_pll/drake_experiment.py]: Agreed -- renamed and commented

File: dair_pll/drake_experiment.py (hunk 5 above)
Comment (mshalm) [reply dair_pll/drake_experiment.py]: Nope! removed.

File: examples/contactnets_simple.py (hunk 14 above)
Comment (mshalm) [reply examples/contactnets_simple.py:114]: I think we need these lines -- unlike the static names, these dynamically pick the data asset to either be the cube or the elbow.

File: examples/contactnets_simple.py (hunk 13 above)
Comment (mshalm) [reply examples/contactnets_simple.py]: Removed.

File: examples/contactnets_simple.py (hunk 14 above)
Comment (ebianchi) [reply examples/contactnets_simple.py:114]: Ah, yes good catch.  Agreed.

File: dair_pll/file_utils.py
Comment (mshalm) [reply dair_pll/file_utils.py:1]: This is a great point and comes from a misunderstanding I had about the current file system organization.

I am now proposing the following modification:

Something I'm trying to capture is that we may run many instances or "runs" of the same _type_ of experiment, and that we don't need to copy the data folder every time necessarily.

So I am proposing the following new folder structure

```
- storage/
  |
  - experiment_name/
    |
    - data/
    - runs/
      |
      - exp_run_name/
        |
        - urdfs/
        - trajectory.gif
        - tensorboard/
      - other_exp_run_name/
        |
        - urdfs/
        - trajectory.gif
        - tensorboard/
```

In this case I'm proposing that, `/..../storage/experiment_name` only needs to be set as `SupervisedLearningExperimentConfig.storage`, and `SupervisedLearningExperimentConfig.run_name` should be set to `exp_run_name`. Everything else would be inferred from these two variables.

Let me know what you think about this!

Comment (ebianchi) [reply dair_pll/file_utils.py:1]: This sounds great to me!

**Comment by mshalm** [comment]
In response to prior discussion, file management has been further updated to account for studies.

The updated file system concept is the following:

- the top-level storage is associated with a dataset e.g. of trajectories.
- "Experiment runs" are individual executions of an abstract experiment run on the dataset, which has specific hyperparameters associated with it, e.g. the model space/ SystemConfig. Storing of final results are now a capability of the `Experiment` class. A single experiment run at the moment is only understood as a `SupervisedLearningExperiment` or training 1 model on the dataaset.
-  "Studies" are analyses depending on multiple experiments run on a given dataset. `study.py` contains the functionality to decide which experiments should be run, as well as meta-analysis of those experiments, such as hyperparameter optimization and dataset size sweeps.

Accordingly, the folder structure is the following:
```
- storage/
  |
  - {storage_name}/
    |
    - data/
    - runs/
      |
      - {run_name}/
        |
        - urdfs/
        - trajectory.gif
        - statistics.pkl
        - tensorboard/
      - {other_run_name}/
        |
        - urdfs/
        - trajectory.gif
        - statistics.pkl
        - tensorboard/
    - studies/
      |
      - {study_name}/
        |
        - optimal_hyperparameters.json
```

File: dair_pll/drake_experiment.py (hunk 8 above)
Comment (mshalm) [reply dair_pll/drake_experiment.py:160]: Conceptually, the joint_system_summary is a `SystemSummary` is a _supplement_ to the learned system summary that compares the base/ground truth system. The video overlay is currently the only summary info that falls under this category. Other reporting like the learned geometry and training errors are by contrast solely dependent on the learned system and the data.

To clarify this, I have renamed this method to `base_and_learned_comparison_summary`.

File: dair_pll/drake_utils.py (hunk 9 above)
Comment (mshalm) [reply dair_pll/drake_utils.py:282]: Something in the updates to `mypy` since 3.8 caused these lines to fail static type checking. `RigidTransform` is an alias for `RigidTransform_[float]` so the functionality is the same.

File: dair_pll/multibody_learnable_system.py (hunk 10 above)
Comment (mshalm) [reply dair_pll/multibody_learnable_system.py]: Great point. Will move to a helper in `tensor_utils.py`

File: dair_pll/multibody_learnable_system.py (hunk 11 above)
Comment (mshalm) [reply dair_pll/multibody_learnable_system.py:329]: See previous comments on learned system summary vs. base-vs-learned system comparison summary.

File: dair_pll/experiment.py [hunk 15]
```diff
@@ -187,14 +192,16 @@ def __init__(self, config: SupervisedLearningExperimentConfig) -> \
...
         self.space = base_system.space
-        self.data_manager = SystemDataManager(base_system, config.data_config)
+        self.data_manager = SystemDataManager(base_system, config.storage,
+                                              config.data_config)
         self.loss_callback = cast(LossCallbackCallable, self.prediction_loss)
         if config.run_tensorboard:
             self.tensorboard_manager = TensorboardManager(
-                self.data_manager.get_tensorboard_folder())
+                file_utils.tensorboard_dir(self.config.storage,
+                                           self.config.run_name))
```
Comment (ebianchi) [inline dair_pll/experiment.py:204]: An unimportant comment:  in line 203 you access the storage via `self.config.storage`, but line 198 accesses it via `config.storage`.  They're the same since line 194 is `self.config = config` but just wanted to point out the inconsistency since I noticed it.

**Review by ebianchi — APPROVED** [review]
Looks good to me!  One tiny thing pointed out in a line comment.  Totally optional.  PR approved on my end.

File: dair_pll/experiment.py (hunk 15 above)
Comment (mshalm) [reply dair_pll/experiment.py:204]: This is going to be removed soon with the migration to W&B, so I'll choose to leave it be for now.
